File: backend/app/models/schedule.py
"""
班表模組 SQLAlchemy ORM Models
包含：departments（部門表）、staff_members（人員表）、shift_types（班別表）、
      schedules（班表主檔）、schedule_details（班表明細）、schedule_import_logs（匯入紀錄）

本模組為純本地 SQLite 資料庫模組，不對接 Ragic。
"""
import uuid
from datetime import datetime
from sqlalchemy import (
    Column, String, Integer, Boolean, DateTime, Date,
    ForeignKey, Text, JSON, func
)
from app.core.database import Base


def _uuid() -> str:
    return str(uuid.uuid4())


# ─────────────────────────────────────────────────────────────
# 1. 部門表
# ─────────────────────────────────────────────────────────────
class Department(Base):
    __tablename__ = "schedule_departments"

    id          = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    name        = Column(String(100), nullable=False, unique=True, default="", comment="部門名稱")
    remark      = Column(String(200), nullable=False, default="", comment="備註")
    sort_order  = Column(Integer, nullable=False, default=0, comment="排序")
    is_active   = Column(Boolean, nullable=False, default=True, comment="是否啟用")
    is_deleted  = Column(Boolean, nullable=False, default=False, comment="軟刪除")
    created_at  = Column(DateTime, nullable=False, server_default=func.now(), comment="建立時間")
    updated_at  = Column(DateTime, nullable=False, server_default=func.now(),
                         onupdate=func.now(), comment="更新時間")

    def __repr__(self) -> str:
        return f"<Department name={self.name}>"


# ─────────────────────────────────────────────────────────────
# 2. 人員表
# ─────────────────────────────────────────────────────────────
class StaffMember(Base):
    __tablename__ = "schedule_staff_members"

    id              = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    staff_code      = Column(String(50),  nullable=False, default="", comment="人員代碼（可空）")
    name            = Column(String(100), nullable=False, default="", comment="姓名（解析後）")
    source_name     = Column(String(150), nullable=False, default="", comment="Excel 原始姓名")
    department_id   = Column(String(36),  ForeignKey("schedule_departments.id"), nullable=True,
                             comment="部門 FK")
    department_name = Column(String(100), nullable=False, default="", comment="部門名稱快照")
    employment_type = Column(String(20),  nullable=False, default="正職",
                             comment="正職 / PT / 支援人員")
    remark          = Column(String(200), nullable=False, default="", comment="備註（如福群）")
    is_active       = Column(Boolean,     nullable=False, default=True,  comment="是否啟用")
    is_deleted      = Column(Boolean,     nullable=False, default=False, comment="軟刪除")
    created_at      = Column(DateTime,    nullable=False, server_default=func.now(), comment="建立時間")
    updated_at      = Column(DateTime,    nullable=False, server_default=func.now(),
                             onupdate=func.now(), comment="更新時間")

    def __repr__(self) -> str:
        return f"<StaffMember name={self.name} type={self.employment_type}>"


# ─────────────────────────────────────────────────────────────
# 3. 班別表
# ─────────────────────────────────────────────────────────────
class ShiftType(Base):
    __tablename__ = "schedule_shift_types"

    id           = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    code         = Column(String(20),  nullable=False, unique=True, default="", comment="班別代碼，如 Y、N1")
    name         = Column(String(100), nullable=False, default="", comment="班別名稱")
    start_time   = Column(String(5),   nullable=False, default="",  comment="上班時間 HH:MM")
    end_time     = Column(String(5),   nullable=False, default="",  comment="下班時間 HH:MM")
    work_minutes = Column(Integer,     nullable=False, default=480, comment="預設工時分鐘（預設 8hr）")
    is_overnight = Column(Boolean,     nullable=False, default=False, comment="是否跨日")
    color        = Column(String(20),  nullable=False, default="#6b7280", comment="前端顯示顏色 hex")
    is_active    = Column(Boolean,     nullable=False, default=True,  comment="是否啟用")
    is_deleted   = Column(Boolean,     nullable=False, default=False, comment="軟刪除")
    created_at   = Column(DateTime,    nullable=False, server_default=func.now(), comment="建立時間")
    updated_at   = Column(DateTime,    nullable=False, server_default=func.now(),
                          onupdate=func.now(), comment="更新時間")

    def __repr__(self) -> str:
        return f"<ShiftType code={self.code} {self.start_time}-{self.end_time}>"


# ─────────────────────────────────────────────────────────────
# 4. 班表主檔
# ─────────────────────────────────────────────────────────────
class Schedule(Base):
    __tablename__ = "schedules"

    id               = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    schedule_year    = Column(Integer,     nullable=False, default=0,  comment="年（西元）")
    schedule_month   = Column(Integer,     nullable=False, default=0,  comment="月")
    title            = Column(String(200), nullable=False, default="", comment="班表標題")
    source_file_name = Column(String(300), nullable=False, default="", comment="來源檔名")
    import_batch_id  = Column(String(36),  nullable=False, default=_uuid, comment="匯入批次 UUID")
    # raw_summary：保留 Excel 統計欄位（應出勤/實出勤/加班/請假）
    # 格式：{"陳志明": {"應出勤": 22, "實出勤": 21, "超時加班": 2, "請假天數": 1}}
    raw_summary      = Column(JSON,        nullable=True,  comment="Excel 原始統計欄位 JSON")
    # status: draft / imported / confirmed
    status           = Column(String(20),  nullable=False, default="imported", comment="班表狀態")
    is_deleted       = Column(Boolean,     nullable=False, default=False, comment="軟刪除")
    created_at       = Column(DateTime,    nullable=False, server_default=func.now(), comment="建立時間")
    updated_at       = Column(DateTime,    nullable=False, server_default=func.now(),
                              onupdate=func.now(), comment="更新時間")

    def __repr__(self) -> str:
        return f"<Schedule {self.schedule_year}/{self.schedule_month:02d} status={self.status}>"


# ─────────────────────────────────────────────────────────────
# 5. 班表明細
# ─────────────────────────────────────────────────────────────
class ScheduleDetail(Base):
    __tablename__ = "schedule_details"

    id            = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    schedule_id   = Column(String(36), ForeignKey("schedules.id"), nullable=False,
                           comment="班表主檔 FK")
    work_date     = Column(Date,       nullable=False, comment="日期")
    staff_id      = Column(String(36), ForeignKey("schedule_staff_members.id"), nullable=True,
                           comment="人員 FK（nullable，允許先以 staff_name 快照保存）")
    staff_name    = Column(String(150), nullable=False, default="", comment="人員姓名快照")
    shift_code    = Column(String(20),  nullable=False, default="", comment="班別代碼")
    shift_type_id = Column(String(36),  ForeignKey("schedule_shift_types.id"), nullable=True,
                           comment="班別 FK（nullable，未知班別時為 null）")
    start_time    = Column(String(5),   nullable=False, default="", comment="實際上班時間 HH:MM")
    end_time      = Column(String(5),   nullable=False, default="", comment="實際下班時間 HH:MM")
    work_minutes  = Column(Integer,     nullable=False, default=0,  comment="工時分鐘")
    raw_value     = Column(String(50),  nullable=False, default="", comment="Excel 原始值")
    remark        = Column(String(300), nullable=False, default="", comment="備註")
    is_deleted    = Column(Boolean,     nullable=False, default=False, comment="軟刪除")
    created_at    = Column(DateTime,    nullable=False, server_default=func.now(), comment="建立時間")
    updated_at    = Column(DateTime,    nullable=False, server_default=func.now(),
                           onupdate=func.now(), comment="更新時間")

    def __repr__(self) -> str:
        return (
            f"<ScheduleDetail {self.work_date} staff={self.staff_name} "
            f"shift={self.shift_code}>"
        )


# ─────────────────────────────────────────────────────────────
# 6. 匯入紀錄
# ─────────────────────────────────────────────────────────────
class ScheduleImportLog(Base):
    __tablename__ = "schedule_import_logs"

    id                  = Column(String(36), primary_key=True, default=_uuid, comment="主鍵 UUID")
    import_batch_id     = Column(String(36), nullable=False, default="", comment="匯入批次 UUID")
    file_name           = Column(String(300), nullable=False, default="", comment="檔名")
    sheet_name          = Column(String(100), nullable=False, default="", comment="Sheet 名稱")
    schedule_year       = Column(Integer,     nullable=False, default=0,  comment="班表年")
    schedule_month      = Column(Integer,     nullable=False, default=0,  comment="班表月")
    total_rows          = Column(Integer,     nullable=False, default=0,  comment="掃描列數")
    total_details       = Column(Integer,     nullable=False, default=0,  comment="明細總筆數")
    success_count       = Column(Integer,     nullable=False, default=0,  comment="成功筆數")
    warning_count       = Column(Integer,     nullable=False, default=0,  comment="警告筆數")
    error_count         = Column(Integer,     nullable=False, default=0,  comment="錯誤筆數")
    # unknown_shift_codes: ["X1", "Z"]
    unknown_shift_codes = Column(JSON,        nullable=True,  comment="未辨識班別代碼清單")
    # new_staff_names: ["陳志明", "李宗銘"]
    new_staff_names     = Column(JSON,        nullable=True,  comment="新增人員清單")
    message             = Column(Text,        nullable=False, default="", comment="匯入訊息")
    created_at          = Column(DateTime,    nullable=False, server_default=func.now(),
                                 comment="建立時間")

    def __repr__(self) -> str:
        return (
            f"<ScheduleImportLog {self.schedule_year}/{self.schedule_month:02d} "
            f"batch={self.import_batch_id[:8]}>"
        )

File: frontend/src/types/schedule.ts
/**
 * 班表模組 TypeScript 型別定義
 * 對應後端 app/models/schedule.py 與 app/routers/schedule.py
 */

// ── 部門 ────────────────────────────────────────────────────
export interface Department {
  id: string
  name: string
  remark: string
  sort_order: number
  is_active: boolean
}

export interface DepartmentInput {
  name: string
  remark?: string
  sort_order?: number
  is_active?: boolean
}

// ── 班別 ────────────────────────────────────────────────────
export interface ShiftType {
  id: string
  code: string
  name: string
  start_time: string    // HH:MM
  end_time: string      // HH:MM
  work_minutes: number
  is_overnight: boolean
  color: string         // hex
  is_active: boolean
}

export interface ShiftTypeInput {
  code: string
  name: string
  start_time?: string
  end_time?: string
  work_minutes?: number
  is_overnight?: boolean
  color?: string
  is_active?: boolean
}

// ── 人員 ────────────────────────────────────────────────────
export interface StaffMember {
  id: string
  name: string
  source_name: string
  staff_code: string
  department_id: string | null
  department_name: string
  employment_type: string   // 正職 / PT / 支援人員
  remark: string
  is_active: boolean
}

export interface StaffMemberInput {
  name: string
  source_name?: string
  staff_code?: string
  department_id?: string | null
  employment_type?: string
  remark?: string
  is_active?: boolean
}

// ── 班表主檔 ─────────────────────────────────────────────────
export interface Schedule {
  id: string
  schedule_year: number
  schedule_month: number
  title: string
  source_file_name: string
  status: 'draft' | 'imported' | 'confirmed'
  created_at: string
}

// ── 表格式班表（橫向日期 × 縱向人員）─────────────────────────
export interface ScheduleTableHeader {
  day: number
  weekday: string   // 一二三四五六日
}

export interface ScheduleTableCell {
  detail_id: string
  shift_code: string
  color: string
  work_minutes: number
}

export interface ScheduleTableRow {
  staff_id: string | null
  staff_name: string
  employment_type: string
  department_name: string
  cells: Record<number, ScheduleTableCell>   // day → cell
  work_days: number
  work_minutes: number
  raw_summary: Record<string, number | string>
}

export interface ScheduleTableData {
  schedule: Schedule | null
  days_in_month: number
  headers: ScheduleTableHeader[]
  rows: ScheduleTableRow[]
}

// ── 明細列表（逐筆） ─────────────────────────────────────────
export interface ScheduleDetailRow {
  id: string
  work_date: string     // YYYY-MM-DD
  weekday: string
  staff_id: string | null
  staff_name: string
  department_name: string
  employment_type: string
  shift_code: string
  shift_name: string
  shift_color: string
  start_time: string
  end_time: string
  work_minutes: number
  remark: string
  schedule_id: string
}

// ── Excel 匯入結果 ───────────────────────────────────────────
export interface ImportResult {
  schedule_id: string | null
  schedule_year?: number
  schedule_month?: number
  year_month_detected: boolean
  already_exists?: boolean
  total_rows?: number
  total_details?: number
  success_count?: number
  warning_count?: number
  error_count?: number
  unknown_shift_codes?: string[]
  new_staff_names?: string[]
  message: string
  import_batch_id: string
  warnings?: string[]
}

// ── 匯入紀錄 ─────────────────────────────────────────────────
export interface ImportLog {
  id: string
  import_batch_id: string
  file_name: string
  sheet_name: string
  schedule_year: number
  schedule_month: number
  total_details: number
  success_count: number
  warning_count: number
  error_count: number
  unknown_shift_codes: string[] | null
  new_staff_names: string[] | null
  message: string
  created_at: string
}

// ── 統計 ─────────────────────────────────────────────────────
export interface MonthlyStats {
  schedule_id: string
  year: number
  month: number
  total_staff: number
  total_details: number
  person_stats: Record<string, {
    work_days: number
    work_minutes: number
    shifts: Record<string, number>
  }>
  shift_stats: Record<string, number>
  daily_stats: Record<string, Record<string, number>>
}

// ── 篩選參數 ─────────────────────────────────────────────────
export interface ScheduleFilters {
  year?: number
  month?: number
  staff_id?: string
  department_id?: string
  shift_code?: string
}
